File: Top_100_Liked_Questions/101._Symmetric_Tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
    if (!root) {
        return true;
    }

    const queue = [[root.left, root.right]];
    
    while (queue.length > 0) {
        const [L, R] = queue.shift();
        
        if (!L && !R) {
            continue;
        }
        
        if (!L != !R || L.val != R.val) {
            return false;
        }
        
        queue.push([L.left, R.right]);
        queue.push([L.right, R.left]);
    }
    
    return true;
}


var isSymmetric2 = function(root) {
    if (!root) {
        return true;
    }
    
    const helper = (L, R) =>  {
        if (!L && !R) {
            return true;
        }
        
        if (!!L != !!R) {
            return false;
        }
        
        return L.val == R.val && helper(L.left, R.right) && helper(L.right, R.left);
    }
    
    return helper(root.left, root.right);
};

File: Top_100_Liked_Questions/101._Symmetric_Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True

        queue = [(root.left, root.right)]
        while queue:
            L, R = queue.pop(0)
            
            if not L and not R:
                continue

            if bool(L) != bool(R) or L.val != R.val:
                return False
            
            queue.append((L.left, R.right))
            queue.append((L.right, R.left))

        return True
    
    
    def isSymmetric2(self, root: TreeNode) -> bool:
        if not root:
            return True
        
        def helper(L, R):
            if not L and not R:
                return True
            
            if bool(L) != bool(R):
                return False
            
            return L.val == R.val and helper(L.left, R.right) and helper(L.right, R.left)
        
        return helper(root.left, root.right)